src/fraction/approx.rs

```diff
@@ -0,0 +1,541 @@
+                    value_as_ratio: Some(ratio),
+                }
+            }
+
+            GenericFraction::Infinity(Sign::Plus) => SqrtSetup {
+                estimate: SqrtApprox::PlusInf,
+                value_as_ratio: None,
+            },
+
+            GenericFraction::NaN => SqrtSetup {
+                estimate: SqrtApprox::NaN,
+                value_as_ratio: None,
+            },
+
+            something_negative => panic!(
+                "cannot take the square root of a negative number ({})",
+                something_negative
+            ),
+        }
+    }
+}
+
+/// Halves `value` in-place while keeping it in simplest form. This is faster than standard
+/// division.
+fn halve_in_place_pos_rational(ratio: &mut Ratio<BigUint>) {
+    // To mutate the numerator and denominator of the ratio we'll take ownership of both and then
+    // put them back when we're done.
+    let (mut numer, mut denom) = std::mem::take(ratio).into();
+
+    if numer.is_even() {
+        numer /= 2_u32;
+    } else {
+        denom *= 2_u32;
+    }
+
+    *ratio = Ratio::new_raw(numer, denom);
+}
+
+/// Adds two `Ratio`s together without reducing the result to simplest form. This is significantly
+/// faster than using the standard addition operator provided by `num`, and may be used as long as
+/// the result does not need to be in simplest form (e.g. within an algorithm which reduces the
+/// output ratio before returning).
+fn add_ratios_raw(lhs: Ratio<BigUint>, rhs: Ratio<BigUint>) -> Ratio<BigUint> {
+    // Don't bother comparing the denominators because it's highly unlikely that they're equal.
+    // Instead, we just go straight to giving the fractions equal denominators.
+    let (mut lhs_numer, lhs_denom) = lhs.into();
+    let (mut rhs_numer, rhs_denom) = rhs.into();
+
+    let common_denom = lhs_denom.lcm(&rhs_denom);
+
+    let lhs_multiplier = &common_denom / &lhs_denom;
+    let rhs_multiplier = &common_denom / &rhs_denom;
+
+    lhs_numer *= lhs_multiplier;
+    rhs_numer *= rhs_multiplier;
+
+    Ratio::new_raw(lhs_numer + rhs_numer, common_denom)
+}
+
+/// Various square root operations for `GenericFraction`.
+impl<T: Clone + Integer + ToBigUint + ToBigInt + GenericInteger> GenericFraction<T> {
+    /// Returns an unsimplified rational approximation of the square root of `self`.
+    ///
+    /// If you need the result to be simplified, use `sqrt_with_accuracy` instead.
+    ///
+    /// The square of the resulting value is guaranteed to be equal to `self` within the bounds of
+    /// `accuracy`. See [`Accuracy`] for more details.
+    ///
+    /// # Panics
+    /// This method will panic if `self` is negative.
+    pub fn sqrt_with_accuracy_raw(&self, accuracy: impl Borrow<Accuracy>) -> SqrtApprox {
+        let accuracy = accuracy.borrow();
+
+        let SqrtSetup {
+            estimate: initial_estimate,
+            value_as_ratio,
+        } = SqrtSetup::for_value(self);
```

Comment:
I wouldn't mind a private method to panic, but for a public one it feels like an overkill?
Perhaps we could return an `Option<>` here to play it safe?
Or maybe even fall back to `NaN` if we want to spare some extra bits?

Otherwise, it might be worth making the method _unsafe_ and let it panic, but then create another one which would produce an Option and be _safe_?

Reply:
Good idea.

Since we're working with high-level types here which explicitly encode the sign, I can't see a reasonable way to implement an `_unchecked` (for example) version that would use unsafe code. There's not really any scope for us to invoke undefined behaviour here.

Perhaps we could have something safe like `sqrt_abs`, which would only calculate the square root of the magnitude value of `self`? Then we can just destructure `self` to obtain its magnitude and ignore the sign (since the two are separate). This would be nice because it's not an inherently unsafe operation, and it seems fairly likely that there will be many instances where the programmer knows that `self` is not negative – it would be annoying to have to use unsafe code to avoid the check in situations like that.

Reply:
> Perhaps we could have something safe like sqrt_abs

sounds like a nice workaround. I like it!

> There's not really any scope for us to invoke undefined behaviour here.

Agree. I guess, my main concern is not UB, but panic which might be caused by user input. Panic is not something that the library users would be able to _catch_ outside of the library, so they would have to manually check the sign everywhere to not let the application crash.